feat(adapters): add vanilla PostgreSQL adapter

jakecoppinger opened the pull request:
This is my first contribution to this repo - please let me know if there are things I'm missing/could improve!

All tests are passing.

<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

## ☕️ Reasoning

What changes are being made? What feature/bug is being fixed here?

## 🧢 Checklist

- [ x] Documentation
- [x] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

Fixes: https://github.com/nextauthjs/next-auth/issues/4897

## 📌 Resources

- [Contributing guidelines](./CONTRIBUTING.md)
- [Code of conduct](./CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

jakecoppinger committed:
Add vanilla Postgres adapter for NextAuth using the pg package.

All tests are passing.

davidmfoley commented on packages/adapter-postgres/README.md:
```diff
@@ -0,0 +1,48 @@
+export default NextAuth({
+  // https://next-auth.js.org/configuration/providers
+  providers: [],
+  adapter: PostrgesAdapter(client)
```
typo (`Postrges`)

jakecoppinger committed:
Fix typo in readme

jakecoppinger committed:
Add basic tsconfig

jakecoppinger committed:
Replace this.getUser with functio body as this is undef

jakecoppinger commented:
I'm still unable to solve a bug - `createSession` being called with an `undefined` `userId`: https://github.com/nextauthjs/next-auth/issues/4897

I initially thought I fixed it, but the issue persists. Are you able to reproduce and suggest a fix? :)

balazsorban44 commented:
Just remembered, could you add the adapter to these lists:

For auto-labeling issues:
https://github.com/nextauthjs/next-auth/blob/main/.github/issue-labeler.yml
https://github.com/nextauthjs/next-auth/blame/main/.github/ISSUE_TEMPLATE/3_bug_adapter.yml#L23-L34

For auto-labeling PRs
https://github.com/nextauthjs/next-auth/blob/main/.github/pr-labeler.yml

balazsorban44 commented:
What do you think about calling it `@next-auth/pg-adapter`? :slightly_smiling_face:

jakecoppinger committed:
Change name from @next-auth/postgres-adapter to @next-auth/pg-adapter

jakecoppinger committed:
Fix package folder to be same as name

jakecoppinger committed:
Add package to issue template, labeller and PR labeller

jakecoppinger commented:
Renamed to `@next-auth/pg-adapter` and updated the files you mentioned :) 
I'm still stuck on this issue, any ideas? https://github.com/nextauthjs/next-auth/pull/4933#issuecomment-1186500481

balazsorban44 commented:
@jakecoppinger could you update the lock file so we can run tests in GitHub actions? (just run `pnpm install` and commit the result)

jakecoppinger committed:
Set next-auth version to ^4

Co-authored-by: Balázs Orbán <info@balazsorban.com>

jakecoppinger committed:
Commit pnpm lock change.

craig-bowers commented:
Thanks for sharing your Postgres adapter! I used this as a starting point to set up NextAuth with [pg-promise](https://github.com/vitaly-t/pg-promise) instead of pg and JavaScript instead of TS.

I am running into an issue though. When I use your adapter and try signing into my Google account via [Google Provider](https://next-auth.js.org/providers/google) I'm getting a new database entry into each of my users, accounts, and sessions tables, but the next-auth.session-token cookie is not getting set when I'm redirected back to http://localhost:3000 and when I try using useSession() on the front-end the status says I'm unauthenticated.

When I bypass the adapter it lets me log in just fine and I can use useSession() to see my Google email, image, etc., and it tells me that I'm authenticated. Obviously none of the data makes its way into Postgres with it bypassed though.
` // adapter: PostgresAdapter(db),`

Any idea on what the problem might be? I can share my [...nextauth].js or adapter file if you think it'll help.

jakecoppinger commented:
@craig-bowers I haven't done any testing with anything but a standard email and password unfortunately, so no ideas on my end.

I'm still stuck on an issue (#4897), did you find a way to fix that by any chance?!

bonesoul commented:
found a fix now?

bonesoul commented:
LF this to get merged.

jakecoppinger commented:
I tried to fix my PR based on the suggestion in this comment, but I'm not having any success unfortunately: https://github.com/nextauthjs/next-auth/issues/4897#issuecomment-1363414338

`getSessionAndUser` doesn't appear to be called. I've added logging in a test project to see what gets returned from the function but it isn't being called.

Any debugging help would be appreciated!

jakecoppinger committed:
Return user id when updating user. Fixes #4897.

jakecoppinger commented:
Fixed! Needed to return `id` (using SQL command `RETURNING`) when creating a new user.

jakecoppinger commented:
Oops, meant to close the issue and not the PR.

Issue is now fixed. Rebased and ready for review & merge!

bonesoul commented:
eager now to get this merged :)

LFCavalcanti commented:
Hi,

Any update to merge this PR?

robahtou commented:
Any updates on this? Happy to help get this through the finish line!

ThangHuuVu commented on packages/adapter-pg/src/index.ts:
```diff
@@ -0,0 +1,220 @@
+      }
+      const sql = `
+        UPDATE sessions set
+        "userId" = $2, expires = $3
```
we don't need to update userId since it's never passed to this function from core.

ThangHuuVu reviewed (changes requested):
hi folks, sorry for the delay - I reviewed the PR, it looks good to me - I added a small comment. To get this PR merged, here are a couple of left-overs that we need to address:
1. Publish it as ESM-only and under `@auth/` namespace. Make sure to have correct `package.json` file and `tsconfig.json` file
2. Add documentation as JSDoc. 
3. Rebase & passing tests

You can take reference from other adapters like Kysely, Drizzle, Prisma etc.

cc @jakecoppinger @robahtou let me know if you have any questions!

ThangHuuVu commented:
hi folks, I finished this PR in https://github.com/nextauthjs/next-auth/pull/8679, so I'm closing this one - thanks everyone!
